[[FIX]] Parse nested templates

leebyron opened the pull request:
ES6 template literals are nestable, jshint's parser is capable, but it's lexer isn't. This patch adds the concept of a stack of contexts to the lexer, replacing the previous boolean flag. This context stack pattern is borrowed from the awesome Acorn parser project, where it's used to enable more than just template literals.

Fixes: #2151
Fixes: #2082
Fixes: #1814

leebyron commented:
BTW, expanding the use of context can help solve #2082. Acorn uses the same approach to solve this problem.

leebyron commented:
Just also fixed #2082 in a 2nd commit in this patch, editing the description of the pull req now.

caitp commented on src/lex.js:
```diff
@@ -1026,19 +1053,17 @@ Lexer.prototype = {
-    tokenType = this.inTemplate ? Token.TemplateTail : Token.StringLiteral;
-    this.inTemplate = false;
+    // Final value is either StringLiteral or TemplateTail
+    tokenType = tokenType === Token.TemplateHead ? Token.StringLiteral : Token.TemplateTail;
```
since you're fixing all this, you should also fix this: don't use Token.StringLiteral for no-substitution templates, otherwise they could be treated as valid directives, which is not true.

caitp replied:
test case:

```js
function fn() {
  `use strict`;
  return "\077";
} // should not yield any warnings
```

leebyron replied:
Great point. I'll add that to this PR

caitp commented:
rest of it lgtm, the other thing could be a followup bug

leebyron commented:
If you're comfortable merging before I get to fixing the directive issue, go nuts. I'll still follow up with that.

caitp commented:
I think I will wait, I need to make sure the startLine thing can still work correctly with the stack structure, @jugglinmike or @rwaldron might also want to have a look

leebyron commented:
Just added your suggestion, @caitp. Let me know if that's in line with what you were looking for.

caitp commented on src/lex.js:
```diff
@@ -31,7 +31,13 @@ var Token = {
   TemplateMiddle: 11,
-  TemplateTail: 12
+  TemplateTail: 12,
+  TemplateString: 13
```
nit: `NoSubstTemplate` maybe?

caitp commented on tests/unit/core.js:
```diff
@@ -818,6 +818,21 @@ exports.testES6TemplateLiteralMultiline = function (test) {
+  ];
+
+  TestRun(test)
+    .addError(2, "Expected an assignment or function call and instead saw an expression.")
```
I don't think this is right, there should be no error for this

caitp replied:
if it is going to be an error, the message should definitely be clearer

caitp replied:
Another case we probably want to check is this:

```js
function fn() {
  `${"use strict"}`;
  return "something-strict-mode-only";
}
```

That will pass, but it's a nice test to have anyways

leebyron commented:
Changed to `NoSubstTemplate` and added that second test.

I think the error is expected, you get the same error for this example:

```js
function fn() {
  42;
}
```

So it's not specific to templates.

leebyron commented:
@caitp - I just added another commit to this patch that covers what you were doing in #2142 but maintaining awareness of the context, so you won't need to rebase over this PR.

I took your test cases from that PR verbatim, but added cases for `yield` expressions as well.

leebyron commented:
Anything else you guys would like me to cover here @caitp @rwaldron?

caitp commented:
good enough, can you try to figure out why coverage decreased? might need some extra test cases

caitp commented:
I'm guessing it's probably for the `yield` multi-line expressions

caitp commented:
hmm nope, I'm wrong about that :<

caitp commented:
oh, you can remove the quotmark handling for backticks, since templates aren't string tokens anymore

caitp commented:
missing a case for [this](https://coveralls.io/builds/1872194/source?filename=src%2Flex.js#L248) as well

leebyron commented:
Good catch, I'll do that. 

From looking at coverall's report, it's hard to tell what didn't get tested. It thinks coverage of style.js went down, but I didn't touch that file. Lex.js also went down but I can dig into that.

caitp commented:
Stuff in the last commit is okay, but I was referring to [this(]https://github.com/jshint/jshint/blob/8e680e6accfac7c52933b4bc20a8b25c8382171a/src/style.js#L66-L70) << block is never hit because templates are not strings

leebyron commented:
I just saw that too, and updated that commit :)

caitp commented:
coverage still seems to have gone down a tiny bit but I think everything relevant to this PR is covered... will merge

leebyron commented:
Thanks @caitp!
